Update Source.java and associate files to include more fields

DoItTomorrow opened the pull request:
Source.java now has a compulsory author field.
Source.java can now accept BiblioFields on construction.
Current implementation is by overloading Source constructor.

#75 #123

Commit:
Expanded Source fields
Changed Source to no require Author as a compulsory field. Tags can now have a type in addition to their name.

Commit:
Updated BiblioField implemenation
BiblioField is now implemented as a seperated object from Tags

DoItTomorrow committed:
Delete Source.java

DoItTomorrow committed:
Changed Source.java to resolve merge conflict

DoItTomorrow committed:
Merge pull request #4 from CS2103-AY1819S2-W13-3/master

Merge from 13-3 Main

Commit:
Modified Source.java to accept BiblioFields without breaking existing code

DoItTomorrow committed:
Merge pull request #5 from CS2103-AY1819S2-W13-3/master

Update merge from 13-3 Main

Commit:
Updates source to include compulsory Author field
Minor textual edits

Commit:
Adds blank line at end of BiblioField.java

Commit:
Checkstype edits for BiblioField and Source

Commit:
Checkstyle change for BiblioField.java

suriruhani reviewed (commented):
Can you please specify what commands are affected by this update? For example I now see `search` would need to be able to take in and search by an `author` argument.

@DoItTomorrow

case141 commented on src/main/java/seedu/address/model/source/Source.java:
```diff
@@ -21,17 +22,31 @@
     private final Detail detail;
 
     private final Set<Tag> tags = new HashSet<>();
+    private final Set<BiblioField> biblioFields = new HashSet<>();
```
Why is a set necessary? According to what was discussed before, shouldn’t a source have only a single bibliography?

DoItTomorrow replied:
Each biblioField represents one field, such as "Name of website" etc as shown in MS word dialogue box below. I implemented it this way to match tag implementation. That way add, search and edit commands can use existing code. I can always just implement it as a single string array, but our whole project is implemented it this way, with discrete objects for everything.

![image](https://user-images.githubusercontent.com/43004683/55787446-db51e280-5ae8-11e9-9cf5-15b5a71aa538.png)

DoItTomorrow replied:
Upon clarification, the issue seems to be that a Source can have multiple BiblioFields of the same type, which should not be the case. I'll resolve that by tonight.

DoItTomorrow commented:
> Can you please specify what commands are affected by this update? For example I now see `search` would need to be able to take in and search by an `author` argument.

Author is implemented it the same way as Title. As it stands, your existing command would not search for anything in the Author field but it should still work fine. If you want to include it your search, you can reuse the code you use for other fields.

BiblioFields are implemented similar to Tags, except each BiblioField contains 2 strings, .type and .body. You can similar reuse any code for Tag searching for searching through BiblioFields.

suriruhani committed:
Merge branch 'master' into master

suriruhani commented:
> > Can you please specify what commands are affected by this update? For example I now see `search` would need to be able to take in and search by an `author` argument.
> 
> Author is implemented it the same way as Title. As it stands, your existing command would not search for anything in the Author field but it should still work fine. If you want to include it your search, you can reuse the code you use for other fields.
> 
> BiblioFields are implemented similar to Tags, except each BiblioField contains 2 strings, .type and .body. You can similar reuse any code for Tag searching for searching through BiblioFields.

Ok, noted. let us merge this soon so the affected commands have time to incorporate the changes.

Commit:
Updated BiblioFields implementation
All biblio fields are now stored in a single BiblioFields object.
There can now only be one copy of each biblio field.
e.g. A source can no longer have two URLs are the same time.

Commit:
Merge branch 'master' of https://github.com/DoItTomorrow/main

Commit:
Checkstyle edits to BiblioFields.java

Commit:
Undid changes to hash function in Source.java

Commit:
Removed edits to toString in Source.java

Commit:
Removed edits to toString in Source.java

Commit:
Reverted changes to equals() in Source.java

ToonDragon reviewed (approved):
Added bliblo fields and author class.  Does not touch the existing implementation since there is still no real interaction with the model itself.
